File: telemetry/aitf/sdk/go/ocsf/claude_compliance_client.go
// Minimal, dependency-free poller for the Claude Compliance Activity Feed.
//
// Pages through GET /v1/compliance/activities using the documented cursor
// contract (after_id / last_id / has_more) with only the Go standard library,
// invoking a handler for each raw Activity record. Pair with
// ClaudeComplianceMapper to normalize the records to OCSF.
//
// Docs: https://platform.claude.com/docs/en/manage-claude/compliance-activity-feed
//
// This mirrors the Python iter_activities poller in
// integrations/anthropic/compliance/client.py.
package ocsf

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
)

// claudeComplianceBaseURL is the default Activity Feed endpoint.
const claudeComplianceBaseURL = "https://api.anthropic.com/v1/compliance/activities"

// claudeComplianceMaxLimit is the maximum page size accepted by the feed.
const claudeComplianceMaxLimit = 5000

// ActivityFeedOptions configures the Claude Compliance Activity Feed poller.
//
// Repeatable filters (ActivityTypes, OrganizationIDs, ActorIDs) are encoded
// with array-bracket query keys (e.g. activity_types[]). AfterID lets a caller
// resume from a previously persisted cursor.
type ActivityFeedOptions struct {
	ActivityTypes   []string
	OrganizationIDs []string
	ActorIDs        []string
	CreatedAtGte    string
	CreatedAtLt     string
	AfterID         string
	Limit           int          // default 100; valid 1..5000
	BaseURL         string       // default https://api.anthropic.com/v1/compliance/activities
	HTTPClient      *http.Client // optional; default http.DefaultClient
}

// activityFeedPage is the decoded JSON envelope for a single feed page.
type activityFeedPage struct {
	Data    []map[string]interface{} `json:"data"`
	HasMore bool                     `json:"has_more"`
	LastID  string                   `json:"last_id"`
}

// IterActivities pages the feed, invoking handler for each Activity. It stops
// when has_more is false, the handler returns an error, or ctx is cancelled.
func IterActivities(ctx context.Context, apiKey string, opts ActivityFeedOptions,
	handler func(activity map[string]interface{}) error) error {
	limit := opts.Limit
	if limit == 0 {
		limit = 100
	}
	if limit < 1 || limit > claudeComplianceMaxLimit {
		return fmt.Errorf("limit must be between 1 and %d", claudeComplianceMaxLimit)
	}

	baseURL := opts.BaseURL
	if baseURL == "" {
		baseURL = claudeComplianceBaseURL
	}
	client := opts.HTTPClient
	if client == nil {
		client = http.DefaultClient
	}

	baseParams := url.Values{}
	baseParams.Set("limit", strconv.Itoa(limit))
	for _, v := range opts.ActivityTypes {
		baseParams.Add("activity_types[]", v)
	}
	for _, v := range opts.OrganizationIDs {
		baseParams.Add("organization_ids[]", v)
	}
	for _, v := range opts.ActorIDs {
		baseParams.Add("actor_ids[]", v)
	}
	if opts.CreatedAtGte != "" {
		baseParams.Set("created_at.gte", opts.CreatedAtGte)
	}
	if opts.CreatedAtLt != "" {
		baseParams.Set("created_at.lt", opts.CreatedAtLt)
	}

	cursor := opts.AfterID
	for {
		params := cloneValues(baseParams)
		if cursor != "" {
			params.Set("after_id", cursor)
		}

		reqURL := baseURL + "?" + params.Encode()
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
		if err != nil {
			return err
		}
		req.Header.Set("x-api-key", apiKey)

		resp, err := client.Do(req)
		if err != nil {
			return err
		}

		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			return err
		}
		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			return fmt.Errorf("compliance activities request failed: status %d: %s", resp.StatusCode, string(body))
		}

		var page activityFeedPage
		if err := json.Unmarshal(body, &page); err != nil {
			return err
		}

		for _, activity := range page.Data {
			if err := handler(activity); err != nil {
				return err
			}
		}

		if !page.HasMore {
			break
		}
		cursor = page.LastID
		if cursor == "" {
			break
		}
	}
	return nil
}

// cloneValues returns a shallow copy of v so per-page mutation (after_id) does
// not leak into the shared base params.
func cloneValues(v url.Values) url.Values {
	out := make(url.Values, len(v))
	for k, vals := range v {
		cp := make([]string, len(vals))
		copy(cp, vals)
		out[k] = cp
	}
	return out
}

File: telemetry/aitf/sdk/rust/src/ocsf/claude_compliance_client.rs
//! Claude Compliance Activity Feed poller (feature `client`).
//!
//! Pages through `GET /v1/compliance/activities` using the documented cursor
//! contract (`after_id` / `last_id` / `has_more`), invoking a handler for each
//! raw Activity record. Pair with [`ClaudeComplianceMapper`] to normalize the
//! records to OCSF. Ported from the Go `ocsf/claude_compliance_client.go`.
//!
//! Docs: <https://platform.claude.com/docs/en/manage-claude/compliance-activity-feed>

use serde::Deserialize;
use serde_json::Value;

use crate::ocsf::claude_compliance::ClaudeComplianceMapper;
use crate::ocsf::AIBaseEvent;

/// The default Activity Feed endpoint.
pub const CLAUDE_COMPLIANCE_BASE_URL: &str = "https://api.anthropic.com/v1/compliance/activities";

/// The maximum page size the feed accepts.
pub const CLAUDE_COMPLIANCE_MAX_LIMIT: i64 = 5000;

/// Options for the Claude Compliance Activity Feed poller.
///
/// Repeatable filters (`activity_types`, `organization_ids`, `actor_ids`) are
/// encoded with array-bracket query keys (e.g. `activity_types[]`). `after_id`
/// lets a caller resume from a previously persisted cursor.
#[derive(Debug, Clone, Default)]
pub struct ActivityFeedOptions {
    pub activity_types: Vec<String>,
    pub organization_ids: Vec<String>,
    pub actor_ids: Vec<String>,
    pub created_at_gte: Option<String>,
    pub created_at_lt: Option<String>,
    pub after_id: Option<String>,
    /// Page size; default 100, valid range 1..=5000.
    pub limit: Option<i64>,
    /// Override the base URL (e.g. for a test server).
    pub base_url: Option<String>,
}

/// The decoded JSON envelope for a single feed page.
#[derive(Debug, Deserialize)]
struct ActivityFeedPage {
    #[serde(default)]
    data: Vec<Value>,
    #[serde(default)]
    has_more: bool,
    #[serde(default)]
    last_id: String,
}

/// Errors raised by the poller.
#[derive(Debug)]
pub enum Error {
    /// `limit` was outside the valid 1..=5000 range.
    InvalidLimit(i64),
    /// The HTTP transport failed.
    Http(String),
    /// A non-2xx HTTP status was returned.
    Status(u16, String),
    /// A response body could not be decoded.
    Decode(String),
    /// A user-supplied handler returned an error.
    Handler(String),
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Error::InvalidLimit(l) => {
                write!(f, "limit must be between 1 and {CLAUDE_COMPLIANCE_MAX_LIMIT}, got {l}")
            }
            Error::Http(e) => write!(f, "http error: {e}"),
            Error::Status(code, body) => {
                write!(f, "compliance activities request failed: status {code}: {body}")
            }
            Error::Decode(e) => write!(f, "decode error: {e}"),
            Error::Handler(e) => write!(f, "handler error: {e}"),
        }
    }
}

impl std::error::Error for Error {}

/// Resolves and validates the effective limit (default 100, valid 1..=5000).
pub fn resolve_limit(limit: Option<i64>) -> Result<i64, Error> {
    let limit = limit.unwrap_or(100);
    if !(1..=CLAUDE_COMPLIANCE_MAX_LIMIT).contains(&limit) {
        return Err(Error::InvalidLimit(limit));
    }
    Ok(limit)
}

/// Builds the static (non-cursor) query parameters for a request, mirroring the
/// Go poller's array-bracket encoding. Returned as `(key, value)` pairs.
pub fn build_base_params(opts: &ActivityFeedOptions, limit: i64) -> Vec<(String, String)> {
    let mut params: Vec<(String, String)> = Vec::new();
    params.push(("limit".to_string(), limit.to_string()));
    for v in &opts.activity_types {
        params.push(("activity_types[]".to_string(), v.clone()));
    }
    for v in &opts.organization_ids {
        params.push(("organization_ids[]".to_string(), v.clone()));
    }
    for v in &opts.actor_ids {
        params.push(("actor_ids[]".to_string(), v.clone()));
    }
    if let Some(gte) = opts.created_at_gte.as_deref().filter(|s| !s.is_empty()) {
        params.push(("created_at.gte".to_string(), gte.to_string()));
    }
    if let Some(lt) = opts.created_at_lt.as_deref().filter(|s| !s.is_empty()) {
        params.push(("created_at.lt".to_string(), lt.to_string()));
    }
    params
}

fn url_encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(b as char)
            }
            _ => out.push_str(&format!("%{b:02X}")),
        }
    }
    out
}

fn encode_query(params: &[(String, String)]) -> String {
    params
        .iter()
        .map(|(k, v)| format!("{}={}", url_encode(k), url_encode(v)))
        .collect::<Vec<_>>()
        .join("&")
}

/// Pages the feed, invoking `handler` for each raw Activity record. Stops when
/// `has_more` is false, the handler returns an error, or a request fails.
pub fn for_each_activity<F>(
    api_key: &str,
    opts: &ActivityFeedOptions,
    mut handler: F,
) -> Result<(), Error>
where
    F: FnMut(Value) -> Result<(), String>,
{
    let limit = resolve_limit(opts.limit)?;
    let base_url = opts
        .base_url
        .clone()
        .unwrap_or_else(|| CLAUDE_COMPLIANCE_BASE_URL.to_string());
    let base_params = build_base_params(opts, limit);

    let mut cursor = opts.after_id.clone().unwrap_or_default();
    loop {
        let mut params = base_params.clone();
        if !cursor.is_empty() {
            params.push(("after_id".to_string(), cursor.clone()));
        }
        let req_url = format!("{base_url}?{}", encode_query(&params));

        let resp = ureq::get(&req_url)
            .set("x-api-key", api_key)
            .call();

        let response = match resp {
            Ok(r) => r,
            Err(ureq::Error::Status(code, r)) => {
                let body = r.into_string().unwrap_or_default();
                return Err(Error::Status(code, body));
            }
            Err(e) => return Err(Error::Http(e.to_string())),
        };

        let body = response
            .into_string()
            .map_err(|e| Error::Http(e.to_string()))?;
        let page: ActivityFeedPage =
            serde_json::from_str(&body).map_err(|e| Error::Decode(e.to_string()))?;

        for activity in page.data {
            handler(activity).map_err(Error::Handler)?;
        }

        if !page.has_more {
            break;
        }
        cursor = page.last_id;
        if cursor.is_empty() {
            break;
        }
    }
    Ok(())
}

/// Convenience: pages the feed and returns every Activity mapped to an
/// [`AIBaseEvent`] via [`ClaudeComplianceMapper`].
pub fn collect_activities_as_events(
    api_key: &str,
    opts: &ActivityFeedOptions,
) -> Result<Vec<AIBaseEvent>, Error> {
    let mapper = ClaudeComplianceMapper::new();
    let mut events = Vec::new();
    for_each_activity(api_key, opts, |activity| {
        events.push(mapper.map_activity(&activity));
        Ok(())
    })?;
    Ok(events)
}
